numpy/doc/structured_arrays.py

```diff
@@ -6,231 +6,430 @@
+structured datatypes, for example for interpreting binary blobs of structured
+data. For this purpose numpy includes specialized features such as subarrays
+and nested datatypes, and allows detailed control over the memory layout of the
+structure. Structured datatypes are designed to mimic 'structs' in the C
+language, making them useful for interfacing with C code.
+
+Users wishing simply to manipulate tabular data with labelled columns are
+enouraged to also consider alternative pydata projects such as pandas, xarray
+or DataArray, which provide convenient and efficient relational operations on
+labelled multi-type data through a higher-level interface.
+
+.. _defining-structured-types:
+
+Structured Datatypes
+====================
+
+Structured datatypes can be thought of as a sequence of bytes of a certain
+length (the structure's :term:`itemsize`) which is interpreted as a collection
+of fields, where each field has a name, a datatype, and a byte offset within
+the structure. The datatype of a field may be any numpy datatype including
+other structured datatypes, and it may also be a :term:`sub-array` which
+behaves like an ndarray of a specified shape. The offsets of the fields are
+arbitrary, and fields may even overlap. These offsets are usually determined
+automatically by numpy but can also be manually specified.
+
+Structured Datatype Specification
+----------------------------------
+
+Structured datatypes may be created using the function :func:`numpy.dtype` with
+one of 3 alternative forms of specification which vary in flexibility and
+conciseness. The forms of specification are further documented in the
+:ref:`Data Type Objects <arrays.dtypes.constructing>` reference page, and in
+summary they are:
+
+1. A list of tuples, one tuple per field
+````````````````````````````````````````
+
+Each tuple has the form ``(fieldname, datatype, shape)`` where shape is
+optional. ``fieldname`` is a string, ``datatype`` may be any object convertible
+to a datatype, and shape is a tuple of integers specifying subarray shape.
+
+ >>> x = np.zeros(3, dtype=[('x', 'f4'), ('y', np.float32), ('z', 'f4', (2,2))])
+ >>> x
+ array([(0.0, 0.0, [[0.0, 0.0], [0.0, 0.0]]),
+        (0.0, 0.0, [[0.0, 0.0], [0.0, 0.0]]),
+        (0.0, 0.0, [[0.0, 0.0], [0.0, 0.0]])],
+       dtype=[('x', '>f4'), ('y', '>f4'), ('z', '>f4', (2, 2))])
 
-In these examples, y is a simple float array consisting of the 2nd field
-in the structured type. But, rather than being a copy of the data in the structured
-array, it is a view, i.e., it shares exactly the same memory locations.
-Thus, when we updated this array by doubling its values, the structured
-array shows the corresponding values as doubled as well. Likewise, if one
-changes the structured array, the field view also changes: ::
+If ``fieldname`` is the empty string (``''``) the field will be given a default
+name of the form ``f#``, where ``#`` is the integer index of the field,
+counting from 0 from the left::
```

Comment:
Not a complaint about this PR, but I don't think I like this behaviour:

```
>>> np.dtype([('', 'f4'),('f0', 'i4'),('z', 'i8')])
ValueError: field 'f0' occurs more than once
```

In #9054, I change this in some cases to be "index within the unnamed values". Is that a good thing?